[SYCL][DOC] Make some refactoring for group sorting spec

capatober opened the pull request:
No changes for interfaces. Just some clarifications to avoid misunderstanding.

Following was done:
1. Moved sorting functions before sorters.
2. Added a comment for bubble_sort that this is not a recommended implementation for performance.
3. Added a comment that users shouldn’t create their own Group Helper. `group_with_scratchpad` is enough.
4. Changed the status for the Spec in the `README.md`.
5. Made an entire extension experimental.

Signed-off-by: Fedorov, Andrey <andrey.fedorov@intel.com>

capatober commented:
Thanks to @gmlueck for useful offline comments.

gmlueck commented:
> |`template<typename Group, typename Ptr>
> void operator()(Group g, Ptr first, Ptr last);`
> |Implements a sorting algorithm that calls by `joint_sort`.
> Available only if `sycl::is_group_v<std::decay_t<Group>>` is true.
> `first`, `last` must be the same for all work-items in the group.

Sorry, GitHub does not allow me to comment on lines you didn't change in this PR.  The above is the description of `operator()` on the sorter class.  I think the last sentence about `first` and `last` isn't worded the right way.  This is not a requirement on the application.  Rather, it is a guarantee that the implementation makes.  I think a better wording would be like this:

> The application defines this function to implement a sorting algorithm for `joint_sort`.  The implementation calls this member function once from every work item in the group, passing the same values of `first` and `last` that were passed to `joint_sort`.  The `Group` template parameter is set according to the matching call to `joint_sort`.

For the other overload of `operator()`:

> The application defines this function to implement a sorting algorithm for `sort_over_group`.  The implementation calls this member function once from every work item in the group, passing the work item's value.  The function must return the value corresponding to this work-item's position in the sorted range.  The `Group` template parameter is set according to the matching call to `sort_over_group`.

gmlueck commented:
> ```
>    template<typename T>
>    static constexpr std::size_t
>    memory_required(sycl::memory_scope scope, std::size_t range_size);
>    template<typename T, int dimensions>
>    static constexpr std::size_t
>    memory_required(sycl::memory_scope scope, sycl::range<dimensions> local_range);
> ```

I was wondering about the names of these member functions from `default_sorter` and `radix_sorter`.  Do we think it will be clear to users about which one to use for `joint_sort` vs. `sort_over_group`?  I wonder if it would be clearer to name them `memory_required_joint_sort` and `memory_required_sort_over_group`?  Or maybe `joint_sort_memory_size` and `sort_over_group_memory_size`?

gmlueck commented:
I found the description of "Group Helper" confusing at first.  I think it would help to replace this introduction:

> To pass additional memory to algorithms that don’t have the Sorter parameter SYCL introduces special type: group helper. It encapsulates a group and a memory.
>
> Group helper must have following methods:

With:

> The overloads of `joint_sort` and `sort_over_group` that do not take a `Sorter` parameter implicitly use the default sorter.  Since the default sorter requires the application to allocate some temporary memory, the application must use a `GroupHelper` object to communicate the location of this memory.  A `GroupHelper` object is any object that has the following two public member functions:

I also think the definition of `get_memory` could be clarified.  This is what you have now:

> `/* unspecified */ get_memory() const`
>
> Returns the memory object that represents a memory handled by the group helper object.
> A type of the returned value must be the same as the type of the `default_sorter` 's constructor
> that passes an additional memory to `default_sorter`.

I don't understand why the return type here is unspecified.  Doesn't the return type need to be `sycl::span<std::byte, Extent>`?  Therefore, it seems like the description could be like this:

> `sycl::span<std::byte, Extent> get_memory() const`
>
> Returns the temporary memory object that the default sorter can use.  The `Extent` must be defined according to the value returned from `default_sorter::memory_required`.

gmlueck commented:
One more global comment.  I think we should make this entire extension "experimental" and move all of the APIs into the "ext::oneapi::experimental" namespace. My reasoning is that it is almost impossible to use this API today in a non-experimental way.  All of the overloads of `joint_sort` and `sort_over_group` that do not take a `Sorter` parameter require the application to pass a `GroupHelper`.  However, the only predefined Group Helper is in the experimental namespace.

Moreover, if the application calls one of the `joint_sort` or `sort_over_group` overloads that does take a `Sorter` parameter, it will almost certainly pass one of the predefined sorters.  However, all of the predefined sorters are also in the experimental namespace.

Therefore, I think that almost every application that uses this extension will end up using experimental features, so I don't see any value in calling some parts non-experimental.  I think we should just say the whole thing is experimental until we are ready to make the predefined Group Helper and the predefined sorters non-experimental.

gmlueck reviewed (commented):
Sorry, I should have started my comments by saying that I like the new organization a lot.  I think it's much clearer now that `joint_sort` and `sort_over_group` are explained first.

I have a couple final comments below.

capatober commented on sycl/doc/extensions/GroupAlgorithms/SYCL_INTEL_group_sort.asciidoc:
```diff
@@ -93,7 +191,9 @@ T operator()(Group g, T val);`
 
-Example of custom Sorter:
+Example of custom Sorter is below. This implementation is not recommended as a performant way
+to implement bubble sort. It's just an example of a custom sorter.
```
Application can write a custom merge sorter with some specific optimizations, e.g. copy elements for disjoint sub-sequences. Users can also write the bitonic sorter that takes into account only sequences with 2^n elements. It can be faster than we can do since we need to work with any data sizes, but users can customize it for their specific cases. I'd suggest to have such example as a source file somewhere in "oneAPI Samples" repo to not overcomplicate the Spec.
Will it be better if we just remove the example of a custom sorter here, for the Spec?

gmlueck replied:
I assume that any of these example custom sorters would be significantly longer than the bubble sort we have now?  I agree that it would not be good to have a huge example in the spec.  If there is no short, reasonable example sorter, then maybe it is better not to have an example.

Regardless, I think it would be good to mention the use cases you list here.  Maybe something like:

> Custom sorters are particularly useful when the application knows the data has some special property.  For example, an application could implement a merge sort if it knows the data is already partially sorted.

capatober commented:
@gmlueck, thank you for your comments!

> I think the last sentence about first and last isn't worded the right way. This is not a requirement on the application. Rather, it is a guarantee that the implementation makes. 

Could you please clarify the difference between "requirement on the application" and "guarantee that the implementation makes"? Here we need to say users on the Spec level that they need to pass parameters that are same for all work-items in one work-group or sub-group. Sorry, I couldn't figure out why it's "a guarantee that the implementation makes". By the way, the same wording is for others Group algorithms, e.g. joint_reduce.

> I was wondering about the names of these member functions from default_sorter and radix_sorter. Do we think it will be clear to users about which one to use for joint_sort vs. sort_over_group? I wonder if it would be clearer to name them memory_required_joint_sort and memory_required_sort_over_group? Or maybe joint_sort_memory_size and sort_over_group_memory_size?

It was already discussed [here](https://github.com/intel/llvm/pull/3989#issuecomment-874944253) Names are too long and `sort` is already in sorter names. In user's code it can look like following:
```cpp
default_sorter<Compare>::memory_required_sort_over_group(/*...*/);
```

> I don't understand why the return type here is unspecified. Doesn't the return type need to be sycl::span<std::byte, Extent>? 

The thing is we are not sure that each group helper should contain `Extent` as a template parameter. It's ok for `group_with_scratchpad` class, but we decided not to specify it for the Group Helper family.

> One more global comment. I think we should make this entire extension "experimental" and move all of the APIs into the "ext::oneapi::experimental" namespace. My reasoning is that it is almost impossible to use this API today in a non-experimental way. All of the overloads of joint_sort and sort_over_group that do not take a Sorter parameter require the application to pass a GroupHelper. However, the only predefined Group Helper is in the experimental namespace.

I see your point. It looks like users need to pass any experimental thing when they call non-experimental function. However, `joint_sort` and `sort_over_group` are already implemented and if we move functions to experimental it can be an API breaking change.

gmlueck commented:
> Here we need to say users on the Spec level that they need to pass parameters that are same for all work-items in one work-group or sub-group. Sorry, I couldn't figure out why it's "a guarantee that the implementation makes".

Maybe I'm confused, but I thought it was the *library* that calls the custom sorter's `operator()` function.  Is that right?  The current wording:

> first, last must be the same for all work-items in the group.

makes it sound like the application needs to pass the same values of `first` and `last` to each work-item.  However, the application doesn't call this function at all, so this is not a requirement on the application.  Rather, we want to say that the application can assume that the library does pass the same values of `first` and `last` to each work item.

gmlueck commented:
> It was already discussed here Names are too long and sort is already in sorter names.

OK, I see there was already a discussion about this.  I guess this part of the API is experimental, so we will get feedback from customers about whether it is confusing.

gmlueck commented:
> The thing is we are not sure that each group helper should contain Extent as a template parameter. It's ok for group_with_scratchpad class, but we decided not to specify it for the Group Helper family.

But the spec says this currently:

> A type of the returned value must be the same as the type of the default_sorter 's constructor that passes an additional memory to default_sorter.

There is only one constructor for `default_sorter` and the parameter "that passes an additional memory" has the type `sycl::span<std::byte, Extent>`.  Therefore, it seems like the spec already says that the type of the return value is `sycl::span<std::byte, Extent>`, but it says this in a very non-obvious way.

gmlueck commented:
> I see your point. It looks like users need to pass any experimental thing when they call non-experimental function. However, joint_sort and sort_over_group are already implemented and if we move functions to experimental it can be an API breaking change.

Hmm.  Have we made a DPC++ release with this support yet?

dsobolev-dev commented on sycl/doc/extensions/GroupAlgorithms/SYCL_INTEL_group_sort.asciidoc:
```diff
@@ -55,6 +55,104 @@ Table 1. Values of the `SYCL_EXT_ONEAPI_GROUP_SORT` macro.
+
+_Complexity_: Let `N` be `last - first`. `O(N*log(N)*log(N))` comparisons.
+
+_Constraints_: Only available if `GroupHelper` was created with a work group or sub_group and
```
`work group` and `sub_group` looks a bit inconsistent. What do you think about using `work-group` and `sub-group`?  `work-group` is used in https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/GroupAlgorithms/SYCL_INTEL_group_algorithms.asciidoc
Moreover, sycl spec (https://www.khronos.org/registry/SYCL/specs/sycl-2020/pdf/sycl-2020.pdf) uses sub-group.

capatober replied:
Good catch!

capatober replied:
Done. Aligned with SYCL 2020 rev 3

capatober commented:
> Maybe I'm confused, but I thought it was the library that calls the custom sorter's operator() function. Is that right?

Ah, by library you mean implementation? Then, you're right.

> There is only one constructor for default_sorter and the parameter "that passes an additional memory" has the type sycl::span<std::byte, Extent>. Therefore, it seems like the spec already says that the type of the return value is sycl::span<std::byte, Extent>, but it says this in a very non-obvious way.

Correct. Will think about it.

> Have we made a DPC++ release with this support yet?

As far as I know, no.

capatober commented:
@gmlueck, I've addressed most of your comments.
Regarding `/* unspecified */ get_memory() const`
What if we say following?
> `/* unspecified */ get_memory() const`
Returns the temporary memory object that the default sorter can use.

I'm imaging that `get_memory()` for some Group Helper can return `std::pair<std::byte*, size_t>`. It can be wrapped to `sycl::span` and passed to `default_sorter` ctor.
Another variant is to always return `sycl::span<std::byte, sycl::dynamic_extent>`.
Both variants help to avoid the `Extent` template parameter.

gmlueck commented:
> I'm imaging that `get_memory()` for some Group Helper can return `std::pair<std::byte*, size_t>`. It can be wrapped to `sycl::span` and passed to `default_sorter` ctor.
> Another variant is to always return `sycl::span<std::byte, sycl::dynamic_extent>`.
> Both variants help to avoid the `Extent` template parameter.

Maybe I'm confused about how the GroupHelper works.  I thought it was the implementation of `joint_sort()` and `sort_over_group()` that called the GroupHelper `get_memory()` function.  Is that not the case?

The application doesn't control the implementation of `joint_sort()` or `sort_over_group()`.  These are part of the DPC++ implementation.  When we (Intel) code the implementation of `joint_sort()` and `sort_over_group()` don't we need to know what the return type is for `get_memory()`?  The logic will be completely different if the return type is `std::pair<std::byte*, size_t>` vs. `sycl::span<std::byte, sycl::dynamic_extent>`.

capatober commented:
> I thought it was the implementation of joint_sort() and sort_over_group() that called the GroupHelper get_memory() function. Is that not the case?

You're right. The best variant is to mention explicitly that we require sycl::span here as `default_sorter`. If we change `default_sorter` then we need to change the `get_memory()` return type.

Pennycook commented on sycl/doc/extensions/GroupAlgorithms/SYCL_INTEL_group_sort.asciidoc:
```diff
@@ -246,8 +325,8 @@ the default sorting algorithm defined by the sorter calling by `joint_sort`.
-use `sycl::memory_scope::work_group` to get memory size required for each work_group;
-use `sycl::memory_scope::sub_group` to get memory size required for each sub_group`.
+use `sycl::memory_scope::work_group` to get memory size required for each work-group;
+use `sycl::memory_scope::sub_group` to get memory size required for each sub-group`.
```
Is there an extra backtick at the end of this line?

capatober replied:
Yes... Thanks. It would be great if CI checks could catch that

Pennycook commented on sycl/doc/extensions/GroupAlgorithms/SYCL_INTEL_group_sort.asciidoc:
```diff
@@ -292,8 +371,8 @@ memory_required(sycl::memory_scope scope, std::size_t range_size)`
-use `sycl::memory_scope::work_group` to get memory size required for each work_group;
-use `sycl::memory_scope::sub_group` to get memory size required for each sub_group`.
+use `sycl::memory_scope::work_group` to get memory size required for each work-group;
+use `sycl::memory_scope::sub_group` to get memory size required for each sub-group`.
```
Same question as above.

capatober commented:
@gmlueck, @Pennycook, @dmitriy-sobolev, changes are applied. Feel free to review

gmlueck commented on sycl/doc/extensions/README.md:
```diff
@@ -38,7 +38,7 @@ DPC++ extensions status:
 | [Platform Context](PlatformContext/PlatformContext.adoc) | Proposal | |
 | [SYCL_EXT_ONEAPI_DEVICE_IF](DeviceIf/device_if.asciidoc) | Proposal | |
-| [SYCL_INTEL_group_sort](GroupAlgorithms/SYCL_INTEL_group_sort.asciidoc)                                                     | Proposal                                  | |
+| [SYCL_INTEL_group_sort](GroupAlgorithms/SYCL_INTEL_group_sort.asciidoc)                                                     | Partially supported                                  | |
```
Let's change this to "Partially supported (experimental)".

What is not yet supported?  Can we list the supported / unsupported features in the API document?  Otherwise, how will a user know what they can use?

Note that I do not want to add information about supported / unsupported features here in the README because this file is getting out-of-date w.r.t. to the specs themselves.  I think we will more likely keep the information up-to-date if it is all in the spec, rather than this separate README.

Pennycook replied:
> Let's change this to "Partially supported (experimental)".

Nit: Why not just "experimental"?  Going forward it might be better if we reserve "supported" for features that are implemented and supported in the sense that customers (of the product compiler) can expect paid support.

gmlueck replied:
That's OK too.  There is an existing example of another extension saying "supported (experimental)", which is why I suggested it.  Either way is fine with me as long as it says "experimental".

I expect to reconsider this README later when we reorg the specs.

capatober replied:
Seems we are agree to change the wording to "experimental"

gmlueck replied:
That change to the README looks good, but my question about "partially supported" remains.  Are there features in this spec that are not yet supported?

capatober replied:
Yes. Currently only sorting functions and default sorter is supported. Radix sorter is not implemented yet

capatober commented:
@gmlueck, @Pennycook, feel free to approve if changes are good enough to be merged.

gmlueck commented:
> @gmlueck, @Pennycook, feel free to approve if changes are good enough to be merged.

I thought you were going to add some NOTE to the spec saying which features are not implemented yet?  Someone reading that spec needs some way to know which features are safe to use.

capatober commented:
@gmlueck, I think `Release Notes` for oneAPI/DPC++ releases is a better place to mention what is implemented. In my opinion everything that is related to the specific release should be mentioned in Release Notes. So, users will also know which oneAPI/DPC++ Compiler release supports features form the Spec. It also helps to not mix specification and implementation in one document.
